### PR TITLE
Gonzales 3.2 - Fix no mergeable selectors

### LINKED ISSUE #495: Epic task: Fixing rules to work with the latest version of Gonzales-pe
Keen to upgrade to the latest version of gonzales with the hope that it will help solve many issues popping up.

Since there was a big change between the version we use now and gonzales v3.2.1, a lot of the tests are failing.

We've created a branch on here (`feature/gonzales-3-develop`) which will have any test fixes PR'd into it.

If anyone would like to help out and fix a rule please let me know in here so that we can assign you and so that we don't waste time working on the same issue.

Once a fix has been merged, I'll update the list.

# Broken rules

## helpers [assigned: benthemonkey]
- [x] collectSuffixExtensions - no extensions
- [x] collectSuffixExtensions - BEM example
- [x] collectSuffixExtensions - many parents and children

## bem depth - scss [assigned: benthemonkey]
- [x] [max-depth: 0]
- [x] [max-depth: 1]

## bem depth - sass [assigned: benthemonkey]
- [x] [max-depth: 0]
- [x] [max-depth: 1]

## brace style - sass [assigned: bgriffith]
- [x] [style: 1tbs, allow-single-line: true]
- [x] [style: 1tbs, allow-single-line: false]
- [x] [style: stroustrup, allow-single-line: true]
- [x] [style: stroustrup, allow-single-line: false]
- [x] [style: allman, allow-single-line: true]
- [x] [style: allman, allow-single-line: false]

## class name format - scss [assigned: benthemonkey]
- [x] [convention: hyphenatedlowercase]
- [x] [convention: hyphenatedlowercase with ignore]
- [x] [convention: camelcase]
- [x] [convention: snakecase]
- [x] [convention: strictbem]
- [x] [convention: hyphenatedbem]
- [x] [convention: RegExp ^[_A-Z]+$]
- [x] [convention: RegExp ^[_A-Z]+$], with convention-explanation

## class name format - sass [assigned: benthemonkey]
- [x] [convention: hyphenatedlowercase]
- [x] [convention: hyphenatedlowercase with ignore]
- [x] [convention: camelcase]
- [x] [convention: snakecase]
- [x] [convention: strictbem]
- [x] [convention: hyphenatedbem]
- [x] [convention: RegExp ^[_A-Z]+$]
- [x] [convention: RegExp ^[_A-Z]+$], with convention-explanation

## empty line between blocks - sass
- [x] without comments - [include: true, allow-single-line-rulesets: true]
- [x] without comments - [include: true, allow-single-line-rulesets: false]

## extends before declarations - scss [assigned: bgriffith]
- [x] enforce

## extends before declarations - sass [assigned: bgriffith]
- [x] enforce

## extends before mixins - sass [assigned: bgriffith]
- [x] enforce

## force attribute nesting - scss [assigned: danpurdy]
- [x] [default]

## force attribute nesting - sass [assigned: danpurdy]
- [x] [default]

## force element nesting - scss [assigned: danpurdy]
- [x] [default]

## force element nesting - sass [assigned: danpurdy]
- [x] [default]

## force pseudo nesting - scss [assigned: danpurdy]
- [x] [default]

## force pseudo nesting - sass [assigned: danpurdy]
- [x] [default]

## id name format - scss [assigned: benthemonkey]
- [x] [convention: hyphenatedlowercase]
- [x] [convention: hyphenatedlowercase with ignore]
- [x] [convention: camelcase]
- [x] [convention: snakecase]
- [x] [convention: RegExp ^[_A-Z]+$]
- [x] [convention: RegExp ^[_A-Z]+$], with convention-explanation

## id name format - sass [assigned: benthemonkey]
- [x] [convention: hyphenatedlowercase]
- [x] [convention: hyphenatedlowercase with ignore]
- [x] [convention: camelcase]
- [x] [convention: snakecase]
- [x] [convention: RegExp ^[_A-Z]+$]
- [x] [convention: RegExp ^[_A-Z]+$], with convention-explanation

## indentation - scss
- [x] [size: 2]

## mixin name format - scss [assigned: benthemonkey]
- [x] [convention: hyphenatedlowercase]
- [x] [convention: camelcase]
- [x] [convention: RegExp ^[_A-Z]+$]
- [x] [convention: allow-leading-underscore false]

## mixin name format - sass [assigned: benthemonkey]
- [x] [convention: hyphenatedlowercase]
- [x] [convention: camelcase]
- [x] [convention: RegExp ^[_A-Z]+$]
- [x] [convention: allow-leading-underscore false]

## mixins before declarations - scss [assigned: danpurdy]
- [x] enforce
- [x] [excludes: limited]

## mixins before declarations - sass [assigned: danpurdy]
- [x] enforce 
- [x] [excludes: limited]

## nesting depth - scss
- [x] [max-depth: 2]

## nesting depth - sass
- [x] [max-depth: 2]

## no mergeable selectors - scss [assigned: danpurdy]
- [x] [default]
- [x] [whitelist: div p]

## no mergeable selectors - sass [assigned: danpurdy]
- [x] [default]
- [x] [whitelist: div p]

## no qualifying elements - scss  [assigned: bgriffith]
- [x] [attribute: false, class: false, id: false]
- [x] [attribute: true, class: false, id: false]
- [x] [attribute: false, class: true, id: false]
- [x] [attribute: false, class: false, id: true]
- [x] [attribute: true, class: true, id: false]

## no qualifying elements - sass  [assigned: bgriffith]
- [x] [attribute: false, class: false, id: false]
- [x] [attribute: true, class: false, id: false]
- [x] [attribute: false, class: true, id: false]
- [x] [attribute: false, class: false, id: true]
- [x] [attribute: true, class: true, id: false]

## placeholder in extend - sass [assigned: bgriffith]
- [x] enforce

## placeholder in extend - sass [assigned: bgriffith]
- [x] enforce

## single line per selector - scss  [assigned: bgriffith]
- [x] enforce

## single line per selector - sass  [assigned: bgriffith]
- [x] enforce

## space after bang - sass [assigned: bgriffith]
- [x] [include: false]
- [x] [include: true]

## space after bang - sass [assigned: bgriffith]
- [x] sass - [include: false] 
- [x] sass - [include: true]

## space after comma - sass [assigned: bgriffith]
- [x] [include: false]

## space around operator - scss  [assigned: bgriffith]
- [x] [include: true]
- [x] [include: false]

## space around operator - sass [assigned: bgriffith]
- [x] [include: true]
- [x] [include: false]

## space before brace - scss [assigned: danez]
- [x] [include: true]
- [x] [include: false]

### PR DESCRIPTION
**Pull request by DanPurdy**
Last rule update for gonzales 3.2 incoming! Sorry it took so long to get round to..

Fixes the no-mergeable-selectors rule for gonzales 3.2 and also closes #495 and #339 and #307 while we're at it.

Warning messages stay the same but most of the method for this has changed and attaches itself to the root stylesheet traverse of gonzales so should be easier to extend/fix in the future.

Please note though that `Sass` syntax still has some fairly serious issues, which we get around in this rule by ignoring at rules completely whereas `scss` format will report on them.

All tests now passing!

`DCO 1.1 Signed-off-by: Dan Purdy danjpurdy@gmail.com`

### CODE REVIEW

File: lib/rules/no-mergeable-selectors.js [hunk 1]
```diff
@@ -2,225 +2,290 @@
...
+  else if (val.is('attributeSelector')) {
+    content = constructSubSelector(val, '[', ']', constructSelector);
+  }
 
-    val.forEach( function (attrib) {
-      selector += constructSelector(attrib);
-    });
+  else if (val.is('atkeyword') ) {
+    content = constructSubSelector(val, '@', '', constructSelector);
+
```
Comment (bgriffith) [inline lib/rules/no-mergeable-selectors.js]: Unnecessary whitespace

File: lib/rules/no-mergeable-selectors.js [hunk 2]
```diff
@@ -2,225 +2,290 @@
...
-  else if (val.is('attribute')) {
-    var selector = '[';
+  else if (val.is('attributeSelector')) {
+    content = constructSubSelector(val, '[', ']', constructSelector);
+  }
 
-    val.forEach( function (attrib) {
-      selector += constructSelector(attrib);
-    });
+  else if (val.is('atkeyword') ) {
```
Comment (bgriffith) [inline lib/rules/no-mergeable-selectors.js]: No space after `val.is('at keyword')`

File: lib/rules/no-mergeable-selectors.js [hunk 3]
```diff
@@ -2,225 +2,290 @@
...
+    content = constructSubSelector(val, '@', '', constructSelector);
+
+  }
+
+  else if (val.is('placeholder') ) {
```
Comment (bgriffith) [inline lib/rules/no-mergeable-selectors.js]: No space after `val.is('placeholder')`

File: lib/rules/no-mergeable-selectors.js [hunk 4]
```diff
@@ -2,225 +2,290 @@
...
+    content = constructSubSelector(val, '%', '', constructSelector);
+  }
+
+  else if (val.is('variable') ) {
```
Comment (bgriffith) [inline lib/rules/no-mergeable-selectors.js]: No space after `val.is('variable')`

File: lib/rules/no-mergeable-selectors.js [hunk 5]
```diff
@@ -2,225 +2,290 @@
...
+ * Checks an atRule to see if if it's part of our mergeable at rule list.
+ * It also checks for Sass syntax as gonzales currently has issues with the syntax
+ *
+ * @param {object} node - The current node / atRule part of our selector
+ * @returns {boolean} Whether this atRule should be merged or not
+ */
+var isMergeableAtRule = function (node) {
+  var isMergeable = false;
+  node.forEach(function (item) {
+    // Gonzales has issues with nest levels in media queries :(
```
Comment (bgriffith) [inline lib/rules/no-mergeable-selectors.js]: Should add a `TODO` here so we remember to fix it.

**Comment by DanPurdy** [comment]
Also fixes #559

**Comment by benthemonkey** [comment]
:+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1:

**Pull request merged**